packages/core/src/browser/frontend-connection-status.ts

```diff
@@ -0,0 +1,356 @@
+    }
+
+    protected updateStatus(success: boolean): ConnectionStatus {
+        this.connectionState = this.connectionState.next(success);
+        return this.connectionState;
+    }
+
+    protected fireStatusChange(event: ConnectionStatus) {
+        this.statusChangeEmitter.fire(event);
+    }
+
+    protected checkAlive(): Promise<boolean> {
+        return new Promise<boolean>(resolve => {
+            const handle = (success: boolean) => {
+                return resolve(success);
+            };
+            const xhr = new XMLHttpRequest();
+            xhr.timeout = this.options.requestTimeout;
+            xhr.onreadystatechange = event => {
+                const { readyState, status } = xhr;
+                if (readyState === XMLHttpRequest.DONE) {
+                    handle(status === 200);
+                }
+            };
+            xhr.onerror = () => handle(false);
+            xhr.ontimeout = () => handle(false);
+            xhr.open('GET', `${this.endpointUrl}/alive`);
+            try {
+                xhr.send();
+            } catch {
+                handle(false);
+            }
+        });
+    }
+
+}
+
+/**
+ * Just in case we need to support a bit more sophisticated state transitions than having `online` and `offline`.
+ * For instance, `pending`, `reconnecting`, etc...
+ */
+export class ConnectionStateMachine implements ConnectionStatus {
+
+    private static readonly MAX_HISTORY = 100;
+
+    public readonly health: number;
+
+    constructor(
+        private readonly props: { readonly threshold: number },
+        public readonly state: ConnectionState = ConnectionState.ONLINE,
+        private readonly history: boolean[] = []) {
+
+        if (this.state === ConnectionState.OFFLINE) {
+            this.health = 0;
+        } else {
+            this.health = this.history.length === 0 ? 100 : Math.round((this.history.filter(success => success).length / this.history.length) * 100);
+        }
+    }
+
+    next(success: boolean): ConnectionStateMachine {
+        const newHistory = this.updateHistory(success);
+        // Initial optimism.
+        let online = true;
+        if (newHistory.length > this.props.threshold) {
+            online = newHistory.slice(-this.props.threshold).some(s => s);
+        }
+        // Ideally, we do not switch back to online if we see any `true` items but, let's say, after three consecutive `true`s.
+        return new ConnectionStateMachine(this.props, online ? ConnectionState.ONLINE : ConnectionState.OFFLINE, newHistory);
+    }
+
+    protected updateHistory(success: boolean) {
+        const updated = [...this.history, success];
+        if (updated.length > ConnectionStateMachine.MAX_HISTORY) {
+            updated.shift();
+        }
+        return updated;
+    }
+
+}
+
+@injectable()
+export class ConnectionStatusStatusBarContribution extends DefaultFrontendApplicationContribution {
+
+    private idle: boolean;
```

Comment:
Why don't get it from `idleService.idle`?

Reply:
I wanted to save the state.